Use proper location for taxes when adding products via admin

Konamiman opened the pull request:
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

When the `woocommerce_adjust_non_base_location_prices` is set to always use the customer location to calculate taxes, and "I will enter prices inclusive of tax" setting is set, adding a product to an order as an admin via the order editor will calculate the wrong price, because no customer/order data is passed to `wc_get_price_excluding_tax` and thus the location of the currently logged in user (the shop admin) is used instead of the location of the customer.

This PR fixes that by changing `WC_AJAX::maybe_add_order_item` so that it passes the order to `WC_Abstract_Order->add_product`, which in turn passes it to `wc_get_price_excluding_tax` which then obtains the customer from the order and uses his location to calculate the taxes to be substracted from the price.

Closes #30231.

### How to test the changes in this Pull Request:

1. Add the following snippet to your store: `add_filter( 'woocommerce_adjust_non_base_location_prices', '__return_false' );`
2. Make sure that you have the "Prices entered with tax" option set as "Yes, I will enter prices inclusive of tax" (_WooCommerce - Settings - Tax - Tax options_).
3. Let A be the country where your currently logged in user is based. Create an user based on a different country B.
4. Go to _WooCommerce - Settings - Tax - Standard rates_ and set 20% for country A and 10% for country B.
5. Create a simple product, taxable with standard rates, with a price of 110.
6. Create an order from admin area, set U as the customer.
7. Add the product you just created to the order.

Without the fix, the price that will appear in the line item is 91,67 (110 minus 20% of taxes). With the fix, the price will be 100 (10% of taxes substracted this time).

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Use proper location for taxes when adding products via admin

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
Modify "wc_get_price_excluding_tax" to use order location for rates

The order is optional and must be passed as an "order" key in "$args".
If present, the order customer will be passed to "WC_Tax::get_rates"
in order to calculate the taxes to substract (provided that
"$product->is_taxable() && wc_prices_include_tax()")

Also "WC_Abstract_Order->add_product" and "WC_AJAX::maybe_add_order_item"
are modified to pass the order being processed so that ultimately
"wc_get_price_excluding_tax" uses it.

This prevents the base tax location to be wrongly used instead of the
customer location when the "woocommerce_adjust_non_base_location_prices"
hook is set to return false.

Konamiman committed:
Add unit test for 'wc_get_price_excluding_tax'.

The test verifies that when an order is passed as an argument,
the order customer is passed to 'WC_Tax::get_rates'.

vedanshujain reviewed (commented):
LGTM! Given that this is a very fundamental change, what do you think about adding more tests? I think we can add at least for:

1. `add_product` method when order is passed in $args with customer having different base address.
2. `add_product` method when order is passed in $args with customer having similar base address.
3. `add_product` method when order is passed in $args without customer.
4. ``add_product` method without order.

Maybe some more variations for when fileter is disabled/enabled as well?

Konamiman commented:
@vedanshujain I think that since we are already testing that `wc_get_price_excluding_tax` is using the order it receives, it would be enough to add one test for verifying that `add_product` is passing the order to `wc_get_price_excluding_tax` and that the product is added with whatever price is returned by that function.

Konamiman committed:
Add a unit test for WC_Abstract_Order->add_product

The test verifies that if an order is passed in arguments, that order
is passed in turn to wc_get_price_excluding_tax.

vedanshujain reviewed (approved):
LGTM!
